snowflake: `adbc_ingest` will fail with "double free" segmentation fault if record batch schema is incorrect

### What happened?

If schema of `RecordBatchReader` doesn't match the actual batch columns - adbc driver crashes.
It's also pretty hard to debug why, because the only lead is "double free or corruption (out)".
Needed to run under valgrind to understand what's going on.
For some reason it fails in `go` with a proper exception "index out of bounds" but then it's not propagated to the python code.

### Stack Trace

_No response_

### How can we reproduce the bug?

```python
    schema = pa.schema(fields=[
        pa.field("name1", pa.string()),
        pa.field("name2", pa.string()),
    ])
    data = [
        {"name1": "aaa"},
        {"name1": "bbb"},
    ]
    reader = pa.RecordBatchReader.from_batches(schema, [pa.RecordBatch.from_pylist(data)])
    with c.cursor() as cur:
        cur.adbc_ingest("test2", reader, mode="create_append")
```

### Environment/Setup

Latest

@pkit Can you please share the package version(s) for which this issue occurred, and any other configuration you may have passed to the driver/connection?

I failed to reproduce this using `adbc-driver-snowflake = 1.1.0`. For me it failed with the stack trace I would have expected it to:
```
panic: arrow/array: number of columns/fields mismatch

goroutine 38 [running]:
github.com/apache/arrow/go/v17/arrow/array.NewRecord(0x1400018e480, {0x14000e82010, 0x1, 0x160009760?}, 0x2)
        /Users/runner/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/array/record.go:151 +0x198
github.com/apache/arrow/go/v17/arrow/cdata.ImportCRecordBatchWithSchema(0x14000581f80?, 0x1400018e480)
        /Users/runner/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/cdata/interface.go:131 +0x248
github.com/apache/arrow/go/v17/arrow/cdata.(*nativeCRecordBatchReader).next(0x14000a9a340)
        /Users/runner/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/cdata/cdata.go:997 +0x1bc
github.com/apache/arrow/go/v17/arrow/cdata.(*nativeCRecordBatchReader).Next(0x14000a9a340)
        /Users/runner/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/cdata/cdata.go:956 +0x20
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.readRecords({0x161bd3558, 0x140005d0aa0}, {0x108fb3898, 0x14000a9a340}, 0x14000118840)
        /Users/runner/work/arrow-adbc/arrow-adbc/adbc/go/adbc/driver/snowflake/bulk_ingestion.go:315 +0x78
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.(*statement).ingestStream.func3()
        /Users/runner/work/arrow-adbc/arrow-adbc/adbc/go/adbc/driver/snowflake/bulk_ingestion.go:249 +0x34
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /Users/runner/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 17
        /Users/runner/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0x98
Abort trap: 6
```

--- a/go/bulk_ingestion.go
+++ b/go/bulk_ingestion.go
@@ -333,8 +333,21 @@ func newWriterProps(mem memory.Allocator, opts *ingestOptions) (*parquet.WriterP
 	return parquetProps, arrowProps
 }
 
-func readRecords(ctx context.Context, rdr array.RecordReader, out chan<- arrow.RecordBatch) error {
+func readRecords(ctx context.Context, rdr array.RecordReader, out chan<- arrow.RecordBatch) (err error) {
 	defer close(out)
+	// The bound RecordReader is producer-supplied and may panic from
+	// Next(): arrow-go's C Data import, for example, panics when an
+	// inbound batch's column count does not match the advertised schema.
+	// Convert any such panic into an ADBC error so the caller gets a
+	// diagnosable failure instead of an aborted host process.
+	defer func() {
+		if r := recover(); r != nil {
+			err = adbc.Error{
+				Msg:  fmt.Sprintf("failed to read record batch from bound stream: %v", r),
+				Code: adbc.StatusInvalidArgument,
+			}
+		}
+	}()
 
 	for rdr.Next() {
 		rec := rdr.RecordBatch()

--- a/go/bulk_ingestion_test.go
+++ b/go/bulk_ingestion_test.go
@@ -29,8 +29,10 @@ import (
 	"io"
 	"testing"
 
+	"github.com/apache/arrow-adbc/go/adbc"
 	"github.com/apache/arrow-go/v18/arrow"
 	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/cdata"
 	"github.com/apache/arrow-go/v18/arrow/memory"
 	"github.com/apache/arrow-go/v18/parquet/pqarrow"
 	"github.com/stretchr/testify/assert"
@@ -139,6 +141,86 @@ func TestQualifiedTableName(t *testing.T) {
 			assert.Equal(t, tt.expected, st.qualifiedTableName())
 		})
 	}
+}
+
+// TestReadRecordsRecoversFromSchemaMismatch exercises the panic-to-error
+// contract of readRecords. A producer-supplied RecordReader whose advertised
+// schema disagrees with the batch it yields will cause arrow-go's cdata
+// import to panic inside Next(); readRecords must turn that into an ADBC
+// error rather than letting it abort the host process.
+func TestReadRecordsRecoversFromSchemaMismatch(t *testing.T) {
+	mem := memory.NewCheckedAllocator(memory.DefaultAllocator)
+	defer mem.AssertSize(t, 0)
+
+	// Reader advertises two columns...
+	advertised := arrow.NewSchema([]arrow.Field{
+		{Name: "name1", Type: arrow.BinaryTypes.String, Nullable: true},
+		{Name: "name2", Type: arrow.BinaryTypes.String, Nullable: true},
+	}, nil)
+
+	// ...but the batch it actually produces only contains one.
+	bldr := array.NewStringBuilder(mem)
+	defer bldr.Release()
+	bldr.Append("aaa")
+	bldr.Append("bbb")
+	col := bldr.NewArray()
+	defer col.Release()
+
+	batchSchema := arrow.NewSchema([]arrow.Field{
+		{Name: "name1", Type: arrow.BinaryTypes.String, Nullable: true},
+	}, nil)
+	batch := array.NewRecordBatch(batchSchema, []arrow.Array{col}, 2)
+	defer batch.Release()
+
+	src := &mismatchedReader{advertisedSchema: advertised, batch: batch}
+
+	// Round-trip through the C Data interface so Next() goes through the
+	// cdata import path where the panic on mismatch originates.
+	var cstream cdata.CArrowArrayStream
+	cdata.ExportRecordReader(src, &cstream)
+
+	importedRdr, err := cdata.ImportCRecordReader(&cstream, advertised)
+	require.NoError(t, err)
+	imported, ok := importedRdr.(array.RecordReader)
+	require.True(t, ok, "imported reader does not implement array.RecordReader: %T", importedRdr)
+	defer imported.Release()
+
+	// A regression would manifest as a panic propagating out of Next().
+	out := make(chan arrow.RecordBatch, 1)
+	go func() {
+		for rec := range out {
+			rec.Release()
+		}
+	}()
+	err = readRecords(context.Background(), imported, out)
+
+	require.Error(t, err, "expected a clean error from a mismatched stream")
+	var adbcErr adbc.Error
+	require.ErrorAs(t, err, &adbcErr)
+	assert.Equal(t, adbc.StatusInvalidArgument, adbcErr.Code)
+	assert.Contains(t, adbcErr.Msg, "mismatch")
+}
+
+// mismatchedReader advertises one schema but yields a batch with a different
+// column count. Used to drive the C Data import path through readRecords.
+type mismatchedReader struct {
+	advertisedSchema *arrow.Schema
+	batch            arrow.RecordBatch
+	emitted          bool
+}
+
+func (r *mismatchedReader) Retain()                        {}
+func (r *mismatchedReader) Release()                       {}
+func (r *mismatchedReader) Schema() *arrow.Schema          { return r.advertisedSchema }
+func (r *mismatchedReader) Err() error                     { return nil }
+func (r *mismatchedReader) RecordBatch() arrow.RecordBatch { return r.batch }
+func (r *mismatchedReader) Record() arrow.RecordBatch      { return r.batch }
+func (r *mismatchedReader) Next() bool {
+	if r.emitted {
+		return false
+	}
+	r.emitted = true
+	return true
 }
 
 func makeRec(mem memory.Allocator, nCols, nRows int) arrow.RecordBatch {

